### PR TITLE
[public-api-server] Forward Origin header where provided

### PR DESCRIPTION
**Pull request by geropl**
## Related Issue(s)

Fixes #

## How to test
 - Go to https://gpl-papi-forward.preview.gitpod-dev.com/workspaces
 - open the network tab
 - create an Organization
 - notice all the public-api calls going through

## Release Notes

```release-note
NONE
```

## Build Options:

- [ ] /werft with-github-actions
      Experimental feature to run the build with GitHub Actions (and not in Werft).
- [ ] leeway-no-cache
      leeway-target=components:all
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-ee-license
- [ ] with-slow-database
- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by geropl** [comment]
/werft run

:snail: DB...

File: components/public-api-server/pkg/auth/context.go [hunk 1]
```diff
@@ -25,6 +25,8 @@ const (
 type Token struct {
 	Type  TokenType
 	Value string
+	// Only relevant for CookieTokenType
+	OriginHeader string
```
Comment (easyCZ) [inline components/public-api-server/pkg/auth/context.go:29]: IMO, the Token should not have the Origin defined. It's an unrelated piece of information. If we need, let's pull this out of the request in the connection constructor, but it doesn't need to be here.

File: components/public-api-server/pkg/proxy/conn.go [hunk 2]
```diff
@@ -85,9 +86,8 @@ func NewConnectionPool(address *url.URL, poolSize int) (*ConnectionPool, error)
 		connConstructor: func(token auth.Token) (gitpod.APIInterface, error) {
```
Comment (easyCZ) [inline components/public-api-server/pkg/proxy/conn.go:86]: We can pass the `context` in here, and use it to extract the Origin.

File: components/public-api-server/pkg/auth/context.go (hunk 1 above)
Comment (geropl) [reply components/public-api-server/pkg/auth/context.go:29]: I'm also not super happy with this approach. But it's not unrelated: It's a required piece to decided whether we can trust the session token.

Comment (geropl) [reply components/public-api-server/pkg/auth/context.go:29]: I wondered if we can rename this struct? :thinking:

Comment (easyCZ) [reply components/public-api-server/pkg/auth/context.go:29]: Sure, but it's not an inherent property of the Token, it's a property of the request that we received. I should be able to use the token, without the Origin information and it's still a perfectly valid token.

Comment (easyCZ) [reply components/public-api-server/pkg/auth/context.go:29]: I'd rather see this be made part of the Server Proxy Conn constructor, as that's where it's required - when establishing the connection, not for the credentials passed into that connection.

Comment (easyCZ) [reply components/public-api-server/pkg/auth/context.go:29]: > I wondered if we can rename this struct? 🤔

What would that achieve? It's not intended to capture request information - for that we either use a request, or the context. It just abstracts over Cookie & Bearer credentials.

File: components/public-api-server/pkg/proxy/conn.go (hunk 2 above)
Comment (easyCZ) [reply components/public-api-server/pkg/proxy/conn.go:86]: or pass the origin explicitly as well

File: components/public-api-server/pkg/auth/context.go (hunk 1 above)
Comment (geropl) [reply components/public-api-server/pkg/auth/context.go:29]: That's fair. But having it as "required" parameter in the `NewCookieToken` method is a very nice property, and I'd like to keep that connectedness. It forces clients to think about the token. :slightly_smiling_face: 

To follow your idea: I thought about introduced a wrapping struct "Authentication" (no idea what the name would be?) that we place in Context instead, which contains the "Token" struct. But that introduced even more changes, and I was not sure the benefits are worth it.
Do you have another proposal that keeps the property of keeping a "connection between cookie and origin"?

Comment (easyCZ) [reply components/public-api-server/pkg/auth/context.go:29]: We can do that, but the Cookie, and the Origin header are not really connected. Yes, in browser scenarios you'd expect the Cookie to be sent only for the Origin, for which the Cookie was defined, but that does not mean I will not send a request which has a cookie and an entirely different origin (The Origin, and the Cookie behaviour are a client side behaviour, not a server side one.

If you want to make sure that connections to `server` over the protocol do send the origin, then this should be done in the Websocket connection constructor, ideally at the `gitpod-protocol` constructor level. This then forces you to supply it in the public-api implementation of the conn constructor, but that does not mean the token needs to encode it.

Comment (geropl) [reply components/public-api-server/pkg/auth/context.go:29]: @easyCZ Do you want to take over here? I have to focus on other things, and I feel we're mostly aligned.

I'd just love to not have to remember to get the token+origin header from somewhere and put it somewhere whenever we want to have a connection, but it should be somehow abstracted. E.g. `cookie, origin, err := auth.GetAuthenticationFromContext` _could_ work. Happy to give that a different name/separate structs/multi return/whatever :slightly_smiling_face:  :pray:

Comment (easyCZ) [reply components/public-api-server/pkg/auth/context.go:29]: If you think we have to change the token, then let's make the following change:

1. Token becomes an interface
```
type Token interface {
  Type() TokenType
  Value() string
  Origin() string
}
```
2. There are two implementations of this:
	* `CookieToken`, which has a helper constructor to do `NewCookieToken(cookie, origin)`
	* `BearerToken`, which has a helper constructor to do `NewBearerToken(token)`
3. Places that use the token can use the type to check, and call the interface methods to get the values.

This at least ensures that we're not growing the `Token` too arbitrarily

Comment (easyCZ) [reply components/public-api-server/pkg/auth/context.go:29]: > @easyCZ Do you want to take over here? I have to focus on other things, and I feel we're mostly aligned.

Sure, will push to this PR

Comment (geropl) [reply components/public-api-server/pkg/auth/context.go:29]: It does not have to be the "Token" struct. (discussing async is hard if we [write messages in parallel](https://github.com/gitpod-io/gitpod/pull/16405#discussion_r1106803971) :see_no_evil: )

Basically happy with whatever if we make it easy to do the right thing for clients. :lotus_position:

**Comment by easyCZ** [comment]
@geropl Added the context-style approach. 

1. Middleware always extracts the Origin
2. The cached client constructor always populates it with the extracted value

File: components/public-api-server/pkg/origin/context.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+)
+
+type contextKey int
+
+const (
+	originContextKey contextKey = iota
+)
+
+func ToContext(ctx context.Context, origin string) context.Context {
+	return context.WithValue(ctx, originContextKey, origin)
```
Comment (geropl) [inline components/public-api-server/pkg/origin/context.go:18]: Ah, I like this!

File: components/public-api-server/pkg/origin/middleware.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+func NewInterceptor() *Interceptor {
+	return &Interceptor{}
+}
+
+type Interceptor struct{}
+
+func (i *Interceptor) WrapUnary(next connect.UnaryFunc) connect.UnaryFunc {
+	return connect.UnaryFunc(func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
+		if req.Spec().IsClient {
+			req.Header().Add("Origin", FromContext(ctx))
```
Comment (geropl) [inline components/public-api-server/pkg/origin/middleware.go:22]: Trying to understand what this branch is for: Is this interceptor both usable for servers and clients (e.g., when forwarding upstream)? :thinking:

File: components/public-api-server/pkg/proxy/conn.go [hunk 5]
```diff
@@ -119,15 +121,23 @@ func NewConnectionPool(address *url.URL, poolSize int) (*ConnectionPool, error)
...
 type ConnectionPool struct {
-	connConstructor func(token auth.Token) (gitpod.APIInterface, error)
+	connConstructor func(context.Context, auth.Token) (gitpod.APIInterface, error)
 
 	// cache stores token to connection mapping
 	cache *lru.Cache
 }
 
 func (p *ConnectionPool) Get(ctx context.Context, token auth.Token) (gitpod.APIInterface, error) {
-	cached, found := p.cache.Get(token)
```
Comment (geropl) [inline components/public-api-server/pkg/proxy/conn.go:130]: :100:

**Comment by geropl** [comment]
@easyCZ Thank you, I like it :heavy_check_mark: 

Would you go ahead and approve your own code? :upside_down_face:

File: components/public-api-server/pkg/origin/middleware.go (hunk 4 above)
Comment (easyCZ) [reply components/public-api-server/pkg/origin/middleware.go:22]: Correct. The way the interceptor is implemented is that you could write a single interceptor for both clients, and servers.

Here, we don't focus on the client version too much, but we do "make it work" by honouring the value that was on the context.

**Review by easyCZ — APPROVED** [review]

**Comment by easyCZ** [comment]
> Would you go ahead and approve your own code? 🙃

Care to create a few more "blank check" PRs for me?

**Pull request merged**